ray_on_golem/server/cluster/nodes.py

```diff
@@ -0,0 +1,381 @@
+                activity.parent.parent.data.issuer_id
+            )
+            connection_uri = self._golem_service.get_connection_uri(ip)
+            ssh_proxy_command = self._golem_service.get_ssh_proxy_command(connection_uri)
+
+            work_context = WorkContext(activity)
+
+            self._add_state_log("[4/9] Deploying image...")
+            await self._deploy_activity(work_context, ip, provider_desc)
+
+            self._add_state_log("[5/9] Starting VM container...")
+            await self._start_activity(work_context, ip, provider_desc)
+
+            self._add_state_log("[6/9] Running bootstrap commands...")
+            await self._upload_node_configuration(
+                work_context, ip, self._ssh_public_key_data, provider_desc
+            )
+
+            self._add_state_log("[7/9] Starting ssh service...")
+            await self._start_ssh_server(work_context, ip, provider_desc)
+
+            self._add_state_log("[8/9] Checking SSH connection...")
+            await self.verify_ssh_connection(activity, ip, ssh_proxy_command)
+
+            await self._golem_service.refresh_network_nodes()
+        except Exception as e:
+            logger.exception(f"Creating new activity failed with `%s`", f"{type(e).__name__}: {e}")
+            await self._stop_activity(activity)
+            raise
+
+        self._add_state_log(f"[9/9] Activity ready on provider: {provider_desc}")
+
+        logger.info(f"Creating new activity done on {provider_desc}, {ip=}, {activity=}")
+
+        return activity, ip, ssh_proxy_command
+
+    async def _deploy_activity(self, context: WorkContext, ip: str, provider_desc: str) -> None:
+        activity = context.activity
+        logger.info(f"Deploying image on {provider_desc}, {ip=}, {activity=}")
+
+        deploy_args = {"net": [self._golem_service.get_deploy_args(ip)]}
+        await context.deploy(deploy_args, timeout=timedelta(minutes=5))
+
+    async def _start_activity(self, context: WorkContext, ip: str, provider_desc: str) -> None:
+        activity = context.activity
+        logger.info(f"Starting VM container on {provider_desc}, {ip=}, {activity=}")
+
+        await context.start()
+
+    async def _upload_node_configuration(
+        self,
+        context: WorkContext,
+        ip: str,
+        ssh_public_key_data: str,
+        provider_desc: str,
+    ) -> None:
+        logger.info(f"Running initial commands on {provider_desc}, {ip=}, {context.activity=}")
+
+        hostname = ip.replace(".", "-")
+
+        await context.run("echo 'ON_GOLEM_NETWORK=1' >> /etc/environment")
+        await context.run(f"echo 'NODE_IP={ip}' >> /etc/environment")
+        await context.run(f"hostname '{hostname}'")
+        await context.run(f"echo '{hostname}' > /etc/hostname")
+        await context.run(f"echo '{ip} {hostname}' >> /etc/hosts")
+        await context.run("mkdir -p /root/.ssh")
+        await context.run(f'echo "{ssh_public_key_data}" >> /root/.ssh/authorized_keys')
```

Comment:
this code, and possibly other code in this module, is no longer valid...

please have a look at the current version: https://github.com/golemfactory/ray-on-golem/blob/main/ray_on_golem/server/services/golem/golem.py#L336